proposals/buildfile.md

```diff
@@ -0,0 +1,186 @@
+  - "/youvolume"
+
+exposedPorts:
+  - "123/udp"
+  - "456"
+  - "789/tcp"
+
+user: root
+workingDirectory: "/somewhere"
+entrypoint:
+  - "java"
+  - "-jar"
+cmd:
+  - "myjar.jar"
+
+layers:
+  properties:
+    # file properties applied to all layers
+    filePermissions: "644"
+    directoryPermissions: "755"
+    user: "0"
+    group: "0"
+    timestamp: "0"
+  entries:
+    - name: "scripts and classes"
+      # file properties only applied to this layer "scripts and classes"
+      properties:
+        filePermissions: "333"
+        directoryPermissions: "777"
+        user: "goose"
+        group: "3"
+        timestamp: "2020-06-03T19:31:50+00:00"
+      files:
+        - from: "target/scripts"
+          to: "/app/scripts"
+          # file properties only applied to this copy directive
+          properties:
+            filePermissions: "777"
+          # another copy for the same layer, with includes and excludes
+        - from: "target/classes"
+          to: "/app/classes"
+          excludes:
+            - "**/goose.class"
+            - "**/moose.class"
+          includes:
+            - "**/*.class"
+
+      # another layer, only globally defined file permissions are applied here
+    - name: "other"
+      files:
+        - from: "build/other"
+          to: "/app"
+
+      # a archive layer using a tar, default mediaType
+    - name: "some tar layer"
+      archive: "build/generated.tar"
+
+      # a foreign layer using the optional mediatype for archive layers
+    - name: "some foreign layer"
+      mediaType: "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip"
+      archive: "https://somewhere.com/layer"
+      # should we include size and digest here? I guess we can always ad tings
+```
+
+## Layers
+
+`layers.entries` are a list of `layer` directives
+
+`layer` directives can be `archive` or `file` layers
+
+a `archive` layer consists of 3 parts
+* `name`: the name/description of the layer (metadata)
+* `archive`: a tar file to include as a layer
+* `mediatype`: the mediatype of archive
+
+a `file` layer consists of 3 parts
+* `name`: the name of the layer (metadata)
+* `properties`: on disk metadata for all files in the layer
+* `files`: a list of *copy directies* (see below)
+
+`files` are a list of *copy directives* that consist of 2 required and 3 optional parts
+* `to` *required*: an absolute path to copy the files to on the container
+* `from` *required*: a file, or directory on disk to copy from
+* `includes`: only includes the patterns matched in this parameter
+* `exludes`: excludes all files (higher prescendence than `includes`) matched in this parameter
+* `properties`: FilesProperties to represent on disk metadata for all files in this copy directive
+
+### FileProperties
+
+A list of properties that can be user specified for each file in a file layer
+* filePermissions: An octal representing permissions for files
+* directoryPermissions: An octal representing permissions for directories
```

Comment:
Note that YAML has native support for representing hex and octal numbers  (https://yaml.org/refcard.html)

Reply:
Currently parsing as string in my prototype, but I'll see what the parser allows.

Reply:
Thinking about it, sure we can accept `0777` instead of `777`, but I think we already have parsing logic to handle this. And `777` feels like a better UX to me even if not technically a correct octal match.

Reply:
Hmm... yeah I think the string type is better. 0777 can be confusing, because you can actual do things like `chmod 4755 setCtrls.sh` and in this context the leading 0 could be interpreted as that you want to clear the first digit.